### PR TITLE
Chunk download

### PR DESCRIPTION
**Pull request by horheynm**
# Desctiption
Chunk download if endpoint supports chunked download. 

# Problem
Downloading large files tend to fail

# Solution
Add chunk download, redownload the chunk if it fails less than N times. 

# Design
Use threads to speed up downloads. Download speed is limited by bandwidth and write speed to the disk. Write to disk speed usually differ by usage. so constant write time. 

Any download is a job. If chunk download, we need to combine and delete the chunk files so they are a job also. 
We have two queues.
1. Job queue - run concurrently -- doesnt matter what file gets downloaded
2. Job queues - run only when the previous job_queue finishes -- used for combining chunks

Ex. 
job_queue1 = Queue()
download_job1 = Job(...)
download_job2 = Job(...)
job_queue1.put(download_job1).put(download_job2)

job_queue2 = Queue()
combine_job1 = Job(...)
job_queue2.put(combine_job1)

job_queues = Queue()
job_queues.put(job_queue1)
job_queues.put(job_queue2) # run this queue if the previous job_queue is done to guarantee downloads.

Usage:
```python3
url = "https://download-some-file.com/"
dest_path = "your/local/folder"
max_retries= 3
downloader = Downloader(
    url=url_path, download_path=dest_path, max_retries=num_retries
)
downloader.download()

```

## Code
```python3
stub = "zoo:llama2-7b-ultrachat200k_llama2_pretrain-pruned80"
print(stub)

model = Model(stub)
p = model.onnx_model.path

```

## Output:
```bash
(.venv) george@gpuserver6:~/sparsezoo$ python3 -m scratch.down
zoo:llama2-7b-ultrachat200k_llama2_pretrain-pruned80
Downloading Chunks: 100%|█████████████████████████████████████████████████████████| 5.27G/5.27G [00:47<00:00, 112MB/s]
Combining Chunks: 100%|██████████████████████████████████████████████████████████| 5.27G/5.27G [00:05<00:00, 1.01GB/s]
(.venv) george@gpuserver6:~/sparsezoo$ 

```

## Testing
Basic url mock and donwload count. actual download is covered in e2e

### CODE REVIEW

File: src/sparsezoo/utils/download.py [hunk 1]
```diff
@@ -12,209 +12,385 @@
...
+        self.url = url
+        self.download_path = download_path
+        self.max_retries = max_retries
+        self.file_size: int = self.get_file_size()
+        self.chunk_bytes = chunk_bytes
+        self.job_queues = Queue()
+        self._lock = threading.Lock()
+
+    def is_range_header_supported(self) -> bool:
+        """Check if chuck download is supported"""
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:95]:
```suggestion
        """Check if chunk download is supported"""
```
haha

File: src/sparsezoo/utils/download.py [hunk 2]
```diff
@@ -12,209 +12,385 @@
...
+        response = requests.head(
+            self.url, headers={"Accept-Encoding": "identity"}, allow_redirects=True
+        )
+        if response.headers.get("accept-ranges") is not None:
+            return True
+        return False
+
+    def get_file_size(self):
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:103]:
```suggestion
    def get_file_size(self) -> int:
```

File: src/sparsezoo/utils/download.py [hunk 3]
```diff
@@ -12,209 +12,385 @@
...
+        if file_size > 0:
+            return file_size
+        raise ValueError(f"Invalid download URL: {self.url}")
+
+    def queue_chunk_download_jobs(self) -> None:
+        """
+        Queues two jobs: file chunk download and combining the chunks.
+
+        This method calculates the number of download jobs needed based on the size of
+         the file to be downloadedand the size of each chunk. It creates download jobs
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:118]:
```suggestion
         the file to be downloaded and the size of each chunk. It creates download jobs
```

File: src/sparsezoo/utils/download.py [hunk 4]
```diff
@@ -12,209 +12,385 @@
...
+         for each chunk, where each job is responsible for downloading a specific byte
+         range of the file. After queuing all the chunk download jobs, it queues an
+         additional job to combine these chunks into the final file.
+
+        Each download job is added to a JobQueue with a descrition (shown in the
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:123]:
```suggestion
        Each download job is added to a JobQueue with a description (shown in the
```

File: src/sparsezoo/utils/download.py [hunk 5]
```diff
@@ -12,209 +12,385 @@
...
+         progress bar), and the jobs areconfigured with the necessary parameters,
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:124]:
```suggestion
         progress bar), and the jobs are configured with the necessary parameters,
```

File: src/sparsezoo/utils/download.py [hunk 6]
```diff
@@ -12,209 +12,385 @@
...
+         including the path where the chunk will be saved andthe 'Range' header to
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:125]:
```suggestion
         including the path where the chunk will be saved and the 'Range' header to
```

File: src/sparsezoo/utils/download.py [hunk 7]
```diff
@@ -12,209 +12,385 @@
...
+         download the specified byte range. The method ensures that the byte ranges
+         are calculated correctly to cover the entire file, even if the file size is
+         not a multiple of the chunk size.
+
+        After all chunk download jobs are queued, if at least one chunk download job
+         exists, a final job is queued to combine all downloaded chunks into the final
+         file and deletethe individual chunk files. This combining job is added to a
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:132]:
```suggestion
         file and delete the individual chunk files. This combining job is added to a
```

File: src/sparsezoo/utils/download.py [hunk 8]
```diff
@@ -12,209 +12,385 @@
...
+            - Utilizes the `combine_chunks_and_delete` method as the final step to
+               combine all chunks and clean up.
+
+        Note:
+            This method does not start the download jobs; it only queues them.
+            The jobs need to be executed by a worker or scheduler that processes the
+             queued JobQueues.
+        """
+        download_jobs: Queue[Job] = JobQueue(description="Downloading Chunks")
+        num_download_jobs = self.file_size // self.chunk_bytes + int(
```
Comment (dbogunowicz) [inline src/sparsezoo/utils/download.py:150]: what's the logic here?

**Review by dbogunowicz — APPROVED** [review]

File: src/sparsezoo/utils/download.py (hunk 8 above)
Comment (horheynm) [reply src/sparsezoo/utils/download.py]: It computes the number of download jobs wrt the chunk size

**Review by bfineran — DISMISSED** [review]
looks great @horheynm, very easy to follow. Would be great to add some simple unit tests to confirm the chunking is happening - e2e functionality is definitely thoroughly covered since every other pathway uses download

File: src/sparsezoo/utils/download.py [hunk 9]
```diff
@@ -12,209 +12,385 @@
...
-                unit_scale=True,  # let tqdm to determine the scale in kilo, mega, etc.
-                unit_divisor=1024,  # is used when unit_scale is true
-            )
+        with open(self.download_path, "ab") as combined_file:
+            for file_path in sorted_chunk_files:
+                chunk_path = os.path.join(chunk_directory, file_path)
+                with open(chunk_path, "rb") as infile:
+                    data = infile.read()
+                    combined_file.write(data)
+                    progress_bar.update(len(data))
```
[the remark this answers is not in the document]
Comment (horheynm) [reply src/sparsezoo/utils/download.py:381]: I used ceil, logic is simpler now

[the remark this answers is not in the document]
Comment (horheynm) [reply src/sparsezoo/utils/download.py:381]: No a different progress bar

File: src/sparsezoo/utils/download.py [hunk 10]
```diff
@@ -12,209 +12,384 @@
...
+            job_queue = self.job_queues.get()
+            is_prev_job_queue_success = False
+            with tqdm(
+                total=self.file_size,
+                unit="B",
+                unit_scale=True,
+                desc=job_queue.description,
+                leave=True,
+            ) as progress_bar:
+                with concurrent.futures.ThreadPoolExecutor(
```
[the remark this answers is not in the document]
Comment (horheynm) [reply src/sparsezoo/utils/download.py:266]: we download and combine. 
if just use threads any of them can execute. 
If you use queue. you can block until all jobs are downloaded, then combine

[the remark this answers is not in the document]
Comment (bfineran) [reply src/sparsezoo/utils/download.py:264]: ah yup, good call

**Comment by horheynm** [comment]
> So @horheynm, just so I understand. The logic now computes the amounts of chunks that are to be downloaded, uses concurrency to download all the chunks asynchronously, and the last job is there to "concatenate" all the chunks into a single file right?

yes

File: src/sparsezoo/utils/download.py (hunk 8 above)
Comment (horheynm) [reply src/sparsezoo/utils/download.py]: I used ceil, logic is simpler now

**Review by bfineran — APPROVED** [review]

**Pull request merged**